daemons/daemon-util.in

```diff
@@ -286,7 +286,13 @@ start() {
   local daemonexec=$(_daemon_executable $name)
 
   if use_systemctl; then
+    local onetime_conf="${DATA_DIR}/${name}.onetime.conf"
+    echo "ONETIME_ARGS=$@" > "$onetime_conf"
+
     systemctl start "${name}.service"
+    ret=$?
+
+    rm -f "$onetime_conf"
```

Comment:
This brings up a few questions:

1. Doesn't this need a `systemctl daemon-reload` to pick up new/changed onetime args?

2. How do the onetime args interact with systemd automatic restart?

3. This needs documentation - `git grep onetime` returns nothing, and it's not even clear to me how this would be used.

Reply:
Thanks for the review.

> 1. Doesn't this need a systemctl daemon-reload to pick up new/changed onetime args?

No, it doesn't need.

<details>
  <summary>Debian GNU/Linux buster/sid + ganeti-2.16.0-5 + systemd + this PR patch</summary>

```
# ls /var/lib/ganeti/ganeti-kvmd.onetime.conf
ls: cannot access '/var/lib/ganeti/ganeti-kvmd.onetime.conf': No such file or directory

# ps aux | grep [g]aneti-kvm
root       2807  0.0  0.0 1073756580 7104 ?     Ss   17:26   0:00 /usr/sbin/ganeti-kvmd -f

# /usr/lib/ganeti/2.16/usr/lib/ganeti/daemon-util stop ganeti-kvmd
# /usr/lib/ganeti/2.16/usr/lib/ganeti/daemon-util start ganeti-kvmd -d

# ps aux | grep [g]aneti-kvm
root       2833  0.0  0.0 1073756580 7776 ?     Ss   17:26   0:00 /usr/sbin/ganeti-kvmd -f -d

# ls /var/lib/ganeti/ganeti-kvmd.onetime.conf
ls: cannot access '/var/lib/ganeti/ganeti-kvmd.onetime.conf': No such file or directory
```
</details>

<details>
  <summary>CentOS 7.6.1810 + ganeti-2.16.1-1.el7.x86_64 + systemd + this PR</summary>

```
# ls /var/lib/ganeti/ganeti-kvmd.onetime.conf
ls: cannot access /var/lib/ganeti/ganeti-kvmd.onetime.conf: No such file or directory

# ps aux | grep [g]aneti-kvm
root      2130  0.0  0.0 198412  5772 ?        Ss   17:00   0:00 /usr/sbin/ganeti-kvmd -f

# /usr/lib64/ganeti/2.16/usr/lib64/ganeti/daemon-util stop ganeti-kvmd
# /usr/lib64/ganeti/2.16/usr/lib64/ganeti/daemon-util start ganeti-kvmd -d

# ps aux | grep [g]aneti-kvm
root      2174  0.0  0.0 198412  5764 ?        Ss   17:00   0:00 /usr/sbin/ganeti-kvmd -f -d

# ls /var/lib/ganeti/ganeti-kvmd.onetime.conf
ls: cannot access /var/lib/ganeti/ganeti-kvmd.onetime.conf: No such file or directory
```
</details>

> 2. How do the onetime args interact with systemd automatic restart?

This assumes that it is run from daemon-util.

1. /usr/lib/ganeti/daemon-util start ganeti-kvmd -d
2. echo "ONETIME_ARGS=-d" > /var/lib/ganeti/ganeti-kvmd.onetime.conf (in `daemon-util`)
3. /usr/sbin/ganeti-kvmd -f -d (run by `ganeti-kvmd.service` under systemd)

```
# ganeti-kvmd.service.in
EnvironmentFile = -@SYSCONFDIR@/default/ganeti
EnvironmentFile = -@LOCALSTATEDIR@/lib/ganeti/ganeti-kvmd.onetime.conf
ExecStart = @SBINDIR@/ganeti-kvmd -f $KVMD_ARGS $ONETIME_ARGS
```

> 3. This needs documentation - git grep onetime returns nothing, and it's not even clear to me how this would be used.

The onetime args is for initscript compatibility. daemon-util start a daemon with $@ as an optional argument:

```
# in daemon-uti
start() {
...
  if type -p start-stop-daemon >/dev/null; then
    start-stop-daemon --start --quiet --oknodo \
      --pidfile $pidfile \
      --startas $daemonexec \
      --chuid $usergroup \
      -- $args "$@"
```

1. /usr/lib/ganeti/daemon-util start ganeti-kvmd -d
2. /usr/sbin/ganeti-kvmd -d (run by `daemon-util`)

<details>
  <summary>Debian GNU/Linux buster/sid + ganeti-2.16.0-5 + initscript</summary>

```
# ps aux | grep [g]aneti-kvm
root       1411  0.0  0.0 1073756580 6184 ?     S    17:07   0:00 /usr/sbin/ganeti-kvmd

# /usr/lib/ganeti/daemon-util stop ganeti-kvmd
# /usr/lib/ganeti/daemon-util start ganeti-kvmd -d

# ps aux | grep [g]aneti-kvm
root       1435  0.0  0.0 1073756580 6200 ?     S    17:08   0:00 /usr/sbin/ganeti-kvmd -d
```
</details>

However, in my cluster's commands.log, there were no use cases for which this optional argument was given. So if we don't need this onetime mechanism, I will delete it and push again.

Reply:
Hi, I found a code that requires ONETIME_ARGS in [bootstrap.py#L1012](https://github.com/ganeti/ganeti/blob/master/lib/bootstrap.py#L1012).

`gnt-cluster master-failover` failed without ONETIME_ARGS under systemd.

I fixed the commit log and pushed PR again. Would you please review it again?

Or should ONETIME_ARGS be restricted to ganeti-confd only?

Reply:
Thank you for the explanations, and I'm sorry for forgetting about this PR.

The only thing left that potentially concerns me is about error handling. The entire script is run as `set -e`, so if `systemctl start` fails, then we might not remove the file. I think:

```
echo … > onetime.conf
trap 'rm -f onetime.conf' EXIT
systemctl start …
rm -f onetime.conf
trap - EXIT
```

should be a safer choice. Or `systemctl start || rc = $?`, for example, like it is done below.

However, that can be done as a subsequent improvement, so all good.